doc-design/source/softwareArchitecture.rst

```diff
@@ -1347,13 +1355,15 @@ How to obtain second order derivatives of the event indicator functions :math:`\
 
 QSS performance considerations:
 
-- Without explicit derivative variables for continuous event indicator (zero-crossing) variables, the QSS zero-crossing variable
+- Without explicit derivative variables for continuous event indicator functions, the QSS zero-crossing variable
   cannot accurately track the function of its dependent variables (for which we will have 1\ :sup:`st` and 2\ :sup:`nd` derivatives)
   and thus will have lower accuracy for zero crossings.
   The accuracy of zero crossings is vital not just for solution accuracy but because QSS must accurately predict crossings to
   get robust FMU crossing event detection due to the indirect method QSS must use to try to get the FMU to detect crossings.
   The need to compute numeric 2\ :sup:`nd` derivatives is also a performance hit.
   For these reasons it is strongly encouraged that explicit derivative variables be set up in the FMU for event indicators.
+  [Stuart: To be clear, you want as an output :math:`\dot z(x_c(t), x_d(t), t)` and :math:`\ddot z(x_c(t), x_d(t), t)`?
+  Please clarify to make sure we have the right order specified.]
```

Comment:
Getting :math:`\dot z(x_c(t), x_d(t), t)` would let us do 2nd order QSS cleanly (without numeric differentiation) and should restore the accurate crossing behavior we recently achieved with the prior __zc/__zc_der approach.

Getting :math:`\ddot z(x_c(t), x_d(t), t)` would further help us to 3rd order QSS efficiently but we'd also need 3rd derivatives of state variables.

I'm not sure that it would make much sense for Modelon to add this without some type of automatic differentiation. Doing it with numeric differentiation inside the FMU should be more efficient than doing it from the QSS side (fewer expensive traversals through the FMI API) but the solution accuracy can decay pretty quickly with even 2 levels of ND unless we move to expensive methods. I think this is also best discussed with Modelon before we formulate it into requirements.

Reply:
@DeadParrot : The above comment contradicts
| To do 3rd order QSS methods efficiently we would want 3rd derivatives of state and zero-crossing variables 
I thought for QSSn, you need n-th order derivatives of states d^(n) x/dt^n and n-th order derivatives of zero crossing functions d^(n) z/dt^n, as also stated in Table 5.2.
Do you agree?

Reply:
Sorry for the confusion. My comments were about what we need to add to the current support so that we could get the necessary derivatives using directional derivative calls. To do QSS2 without numeric differentiation we would need to add only 1st derivatives of zero crossing functions as long as those derivative variables could be used in the directional derivative calls. (For state variables we already have derivative variables that we can pass to the directional derivative call to get 2nd derivatives.)

But in the blank slate view, yes, we need a way to get nth derivatives of state and zero crossing variables to do QSSn.